File: scala/AbstractTraits.scala
// can be done in php (with traits) and Java (default method of interface)
object AbstractTraits {
  def main(args: Array[String]){
    val w = new MyHelloWorld(" world!");
    w.sayHelloWorld
  }
}

trait Hello {
  def sayHelloWorld = {
    println("Hello" + getWorld());
  } 
  def getWorld(): String;
}


class MyHelloWorld(world: String) extends Hello {
  def getWorld = world
}

File: php/abstract_traits.php
<?php
// can be done in Java with default interface method
// and also in scala with traits

trait Hello {
  public function sayHelloWorld() {
    echo 'Hello'.$this->getWorld();
  }
  abstract public function getWorld();
}

class MyHelloWorld {
  private $world;
  use Hello;
  public function __construct($world){
    $this->world = $world;
  }

  public function getWorld() {
    return $this->world;
  }
}

$w = new MyHelloWorld(' world!');
$w->sayHelloWorld();




?>

